wave_lang/kernel/wave/multicast.py

```diff
@@ -0,0 +1,165 @@
+# Copyright 2025 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+"""
+Multicast optimization pass.
+
+This pass optimizes tensor loads using cluster-based multicast operations.
+Clusters are defined by the `workgroups_per_cluster` field in HardwareConstraint.
+
+The pass analyzes tensor load patterns and determines if they can benefit from
+multicast operations where a single load operation can be shared across multiple
+workgroups within a cluster.
+"""
+
+import logging
+import math
+import sympy
+
+from .._support.indexing import IndexExpr, IndexSymbol, IndexSequence, safe_subs
+from .._support.tracing import CapturedTrace
+from ..lang.global_symbols import WORKGROUP_0, WORKGROUP_1, WORKGROUP_2
+from ..ops.wave_ops import TensorLoadToLDS, get_custom
+from .compile_options import WaveCompileOptions
+from .constraints import Constraint
+from .utils.general_utils import get_hardware_constraint
+
+logger = logging.getLogger(__name__)
+
+
+def compute_multicast_mask(
+    global_tile_index: dict[IndexSymbol, IndexSequence],
+    workgroups_per_cluster: tuple[int, int, int],
+) -> IndexExpr:
+    """
+    Compute the multicast mask based on global tile access pattern.
+
+    The mask is a 16-bit value where each bit corresponds to a workgroup in the cluster.
+    Workgroups that load the same data will have their corresponding bits set in the mask.
+
+    Args:
+        global_tile_index: Symbolic access pattern for the tensor load
+        workgroups_per_cluster: Cluster dimensions (X, Y, Z)
+
+    Returns:
+        Mask (symbolic expression) indicating which workgroups share this data.
+        Returns 0 if all workgroups load different data (no multicast benefit).
```

Comment:
In the case where each workgroup loads a distinct tile, the mask should be 0, but if you look at the code I don't think this happens. I think a bit is still set in that case. Am I reading this right?